File: server/controllers/environment_controller.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from database import get_db
from services.environment_service import EnvironmentService
from schemas.environment import EnvironmentUpdate, EnvironmentResponse, AdaptationHints

router = APIRouter(prefix="/api", tags=["environment"])


@router.get("/environment", response_model=EnvironmentResponse)
def get_environment(db: Session = Depends(get_db)):
    """Get current environment state."""
    service = EnvironmentService(db)
    env = service.get_environment()
    return env


@router.put("/environment", response_model=EnvironmentResponse)
def update_environment(request: EnvironmentUpdate, db: Session = Depends(get_db)):
    """Update environment state."""
    service = EnvironmentService(db)
    env = service.update_environment(
        temperature=request.temperature,
        humidity=request.humidity,
        wind_speed=request.wind_speed,
        noise=request.noise,
        brightness=request.brightness
    )
    return env


@router.post("/environment/reset", response_model=EnvironmentResponse)
def reset_environment(db: Session = Depends(get_db)):
    """Reset environment to defaults."""
    service = EnvironmentService(db)
    env = service.reset_environment()
    return env


@router.get("/environment/hints", response_model=AdaptationHints)
def get_adaptation_hints(db: Session = Depends(get_db)):
    """Get UI adaptation hints based on environment."""
    service = EnvironmentService(db)
    hints = service.get_adaptation_hints()
    return hints

File: server/controllers/environmentController.js
const EnvironmentService = require('../services/environmentService');

const EnvironmentController = {
  // GET /api/environment
  getEnvironment(req, res) {
    try {
      const environment = EnvironmentService.getEnvironment();
      res.json(environment);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  },

  // PUT /api/environment
  updateEnvironment(req, res) {
    try {
      const { temperature, humidity, wind_speed, noise, brightness } = req.body;

      const environment = EnvironmentService.updateEnvironment({
        temperature,
        humidity,
        wind_speed,
        noise,
        brightness
      });
      res.json(environment);
    } catch (error) {
      if (error.message.includes('must be') || error.message.includes('Invalid')) {
        res.status(400).json({ error: error.message });
      } else {
        res.status(500).json({ error: error.message });
      }
    }
  },

  // POST /api/environment/reset
  resetEnvironment(req, res) {
    try {
      const environment = EnvironmentService.resetEnvironment();
      res.json(environment);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  },

  // GET /api/environment/hints
  getAdaptationHints(req, res) {
    try {
      const hints = EnvironmentService.getAdaptationHints();
      res.json(hints);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  }
};

module.exports = EnvironmentController;
